### PR TITLE
feat(connector_events): added api to fetch connector event logs

### PR DESCRIPTION
**Pull request by ivor-juspay**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
adding GET api for query to fetch connector events log by specified `payment_id`

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
```
curl 'http://localhost:8080/analytics/v1/connector_event_logs?type=Payment&payment_id=pay_yfkCogrWxVbB8O0biPkh' \
  -H 'Accept: */*' \
  -H 'Accept-Language: en-GB,en' \
  -H 'Authorization: Bearer *' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Origin: http://localhost:9000' \
  -H 'Referer: http://localhost:9000/' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-site' \
  -H 'Sec-GPC: 1' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36' \
  -H 'api-key: hyperswitch' \
  -H 'sec-ch-ua: "Chromium";v="112", "Brave";v="112", "Not:A-Brand";v="99"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  --compressed
```

`[{"merchant_id":"Allconnector123","payment_id":"pay_yfkCogrWxVbB8O0biPkh","connector_name":"stripe","request_id":"018cf773-0564-77e8-9e89-6870acfff2db","flow":"Authorize","request":"{\"amount\":6500,\"currency\":\"USD\",\"statement_descriptor_suffix\":null,\"statement_descriptor\":null,\"metadata[order_id]\":\"pay_yfkCogrWxVbB8O0biPkh_1\",\"metadata[is_refund_id_as_reference]\":null,\"return_url\":\"https://sandbox.hyperswitch.io/payments/pay_yfkCogrWxVbB8O0biPkh/Allconnector123/redirect/response/stripe\",\"confirm\":true,\"mandate\":null,\"payment_method\":null,\"customer\":\"*** alloc::string::String ***\",\"description\":\"Hello this is description\",\"shipping[address][city]\":\"Banglore\",\"shipping[address][country]\":\"US\",\"shipping[address][line1]\":\"*** alloc::string::String ***\",\"shipping[address][line2]\":\"*** alloc::string::String ***\",\"shipping[address][postal_code]\":\"*** alloc::string::String ***\",\"shipping[address][state]\":\"*** alloc::string::String ***\",\"shipping[name]\":\"*** alloc::string::String ***\",\"shipping[phone]\":\"*** alloc::string::String ***\",\"payment_method_data[billing_details][email]\":null,\"payment_method_data[billing_details][address][country]\":null,\"payment_method_data[billing_details][name]\":null,\"payment_method_data[billing_details][address][city]\":null,\"payment_method_data[billing_details][address][line1]\":null,\"payment_method_data[billing_details][address][line2]\":null,\"payment_method_data[billing_details][address][postal_code]\":null,\"payment_method_data[type]\":\"card\",\"payment_method_data[card][number]\":\"400000**********\",\"payment_method_data[card][exp_month]\":\"*** alloc::string::String ***\",\"payment_method_data[card][exp_year]\":\"*** alloc::string::String ***\",\"payment_method_data[card][cvc]\":\"*** alloc::string::String ***\",\"payment_method_options[card][request_three_d_secure]\":\"any\",\"capture_method\":\"automatic\",\"payment_method_options\":null,\"setup_future_usage\":null,\"off_session\":null,\"payment_method_types[0]\":\"card\"}","response":"{\\n  \\\"id\\\": \\\"pi_3OXInzD5R7gDAGff1cGAxxHo\\\",\\n  \\\"object\\\": \\\"payment_intent\\\",\\n  \\\"amount\\\": 6500,\\n  \\\"amount_capturable\\\": 0,\\n  \\\"amount_details\\\": {\\n    \\\"tip\\\": {}\\n  },\\n  \\\"amount_received\\\": 0,\\n  \\\"application\\\": null,\\n  \\\"application_fee_amount\\\": null,\\n  \\\"automatic_payment_methods\\\": null,\\n  \\\"canceled_at\\\": null,\\n  \\\"cancellation_reason\\\": null,\\n  \\\"capture_method\\\": \\\"automatic\\\",\\n  \\\"client_secret\\\": \\\"pi_3OXInzD5R7gDAGff1cGAxxHo_secret_E2nuQt3TQz1PbCGT8B6MiWLGu\\\",\\n  \\\"confirmation_method\\\": \\\"automatic\\\",\\n  \\\"created\\\": 1704958559,\\n  \\\"currency\\\": \\\"usd\\\",\\n  \\\"customer\\\": \\\"cus_OuQL3sWtKEn0Cs\\\",\\n  \\\"description\\\": \\\"Hello this is description\\\",\\n  \\\"invoice\\\": null,\\n  \\\"last_payment_error\\\": null,\\n  \\\"latest_charge\\\": null,\\n  \\\"livemode\\\": false,\\n  \\\"metadata\\\": {\\n    \\\"order_id\\\": \\\"pay_yfkCogrWxVbB8O0biPkh_1\\\"\\n  },\\n  \\\"next_action\\\": {\\n    \\\"redirect_to_url\\\": {\\n      \\\"return_url\\\": \\\"https://sandbox.hyperswitch.io/payments/pay_yfkCogrWxVbB8O0biPkh/Allconnector123/redirect/response/stripe\\\",\\n      \\\"url\\\": \\\"https://hooks.stripe.com/redirect/authenticate/src_1OXIo0D5R7gDAGffxkofsikB?client_secret=src_client_secret_1P7sv3nHuAc684vjLZlUhwT1&source_redirect_slug=test_YWNjdF8xTTdmVGFENVI3Z0RBR2ZmLF9QTTBwNkJKN01LOWIzeFA3bkMxNjhCd2tra01oUnZ50100WMdcNcbT\\\"\\n    },\\n    \\\"type\\\": \\\"redirect_to_url\\\"\\n  },\\n  \\\"on_behalf_of\\\": null,\\n  \\\"payment_method\\\": \\\"pm_1OXInzD5R7gDAGffpXWPgsBW\\\",\\n  \\\"payment_method_configuration_details\\\": null,\\n  \\\"payment_method_options\\\": {\\n    \\\"card\\\": {\\n      \\\"installments\\\": null,\\n      \\\"mandate_options\\\": null,\\n      \\\"network\\\": null,\\n      \\\"request_three_d_secure\\\": \\\"any\\\"\\n    }\\n  },\\n  \\\"payment_method_types\\\": [\\n    \\\"card\\\"\\n  ],\\n  \\\"processing\\\": null,\\n  \\\"receipt_email\\\": null,\\n  \\\"review\\\": null,\\n  \\\"setup_future_usage\\\": null,\\n  \\\"shipping\\\": {\\n    \\\"address\\\": {\\n      \\\"city\\\": \\\"Banglore\\\",\\n      \\\"country\\\": \\\"US\\\",\\n      \\\"line1\\\": \\\"sdsdfsdf\\\",\\n      \\\"line2\\\": \\\"hsgdbhd\\\",\\n      \\\"postal_code\\\": \\\"571201\\\",\\n      \\\"state\\\": \\\"zsaasdas\\\"\\n    },\\n    \\\"carrier\\\": null,\\n    \\\"name\\\": \\\"Bopanna MJ\\\",\\n    \\\"phone\\\": \\\"+1123456789\\\",\\n    \\\"tracking_number\\\": null\\n  },\\n  \\\"source\\\": null,\\n  \\\"statement_descriptor\\\": null,\\n  \\\"statement_descriptor_suffix\\\": null,\\n  \\\"status\\\": \\\"requires_action\\\",\\n  \\\"transfer_data\\\": null,\\n  \\\"transfer_group\\\": null\\n}","error":null,"status_code":0,"latency":891,"created_at":"2024-01-11T07:36:00.507Z","method":"POST"}]`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/analytics/src/connector_events/core.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        AnalyticsProvider::Sqlx(_) => Err(FiltersError::NotImplemented(
+            "Connector Events not implemented for SQLX",
+        ))
+        .into_report()
+        .attach_printable("SQL Analytics is not implemented for Connector Events"),
+        AnalyticsProvider::Clickhouse(pool) => get_connector_events(&merchant_id, req, pool).await,
+        AnalyticsProvider::CombinedSqlx(_sqlx_pool, ckh_pool)
+        | AnalyticsProvider::CombinedCkh(_sqlx_pool, ckh_pool) => {
+            get_connector_events(&merchant_id, req, ckh_pool).await
+        }
```
Comment (lsampras) [inline crates/analytics/src/connector_events/core.rs:23]:
```suggestion
        AnalyticsProvider::Clickhouse(pool)
        | AnalyticsProvider::CombinedSqlx(_sqlx_pool, ckh_pool)
        | AnalyticsProvider::CombinedCkh(_sqlx_pool, ckh_pool) => {
            get_connector_events(&merchant_id, req, ckh_pool).await
        }
```

File: crates/analytics/src/connector_events/events.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    //TODO!: update the execute_query function to return reports instead of plain errors...
+    query_builder
+        .execute_query::<ConnectorEventsResult, _>(pool)
+        .await
+        .change_context(FiltersError::QueryBuildingError)?
+        .change_context(FiltersError::QueryExecutionFailure)
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
```
Comment (lsampras) [inline crates/analytics/src/connector_events/events.rs:48]:
```suggestion
#[derive(Debug, serde::Deserialize)]
```

**Review by lsampras — COMMENTED** [review]
can u add a screenshot of postman depicting the request/response as well

File: crates/analytics/src/connector_events/core.rs (hunk 1 above)
Comment (ivor-juspay) [reply crates/analytics/src/connector_events/core.rs]: `  --> crates/analytics/src/connector_events/core.rs:19:9
   |
19 |         AnalyticsProvider::Clickhouse(ckh_pool)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern doesn't bind `_sqlx_pool`
20 |         | AnalyticsProvider::CombinedSqlx(_sqlx_pool, ckh_pool)
   |                                           ---------- variable not in all patterns
21 |         | AnalyticsProvider::CombinedCkh(_sqlx_pool, ckh_pool) => {
   |                                          ---------- variable not in all patterns`

**Comment by ivor-juspay** [comment]
Screenshot 2024-01-11 at 2 19 16 PM

**Review by harsh-sharma-juspay — DISMISSED** [review]

**Review by lsampras — DISMISSED** [review]

**Review by harsh-sharma-juspay — DISMISSED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**